utils/rowgroup/rowstorage.h

```diff
@@ -0,0 +1,329 @@
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+   MA 02110-1301, USA. */
+
+#ifndef ROWSTORAGE_H
+#define ROWSTORAGE_H
+
+#include "rowgroup.h"
+#include <sys/stat.h>
+#include <unistd.h>
+
+namespace rowgroup
+{
+
+class MemManager;
+class RowPosHashStorage;
+class RowGroupStorage;
+
+class RowAggStorage
+{
+public:
+  RowAggStorage(const std::string& tmpDir,
+                RowGroup* rowGroupOut,
+                RowGroup* keysRowGroup,
+                uint32_t keyCount,
+                size_t maxRows = 256,
+                joblist::ResourceManager* rm = nullptr,
+                boost::shared_ptr<int64_t> sessLimit = {},
+                bool enabledDiskAgg = false,
+                bool allowGenerations = false);
+
+  RowAggStorage(const std::string& tmpDir,
+                RowGroup* rowGroupOut,
+                uint32_t keyCount,
+                size_t maxRows = 256,
+                joblist::ResourceManager* rm = nullptr,
+                boost::shared_ptr<int64_t> sessLimit = {},
+                bool enabledDiskAgg = false,
+                bool allowGenerations = false)
+      : RowAggStorage(tmpDir, rowGroupOut, rowGroupOut, keyCount, maxRows,
+                      rm, std::move(sessLimit),
+                      enabledDiskAgg, allowGenerations)
+  {}
+
+  ~RowAggStorage();
+
+  /** @brief Find or create resulting row.
+   *
+   *    Create "aggregation key" row if necessary.
+   *    NB! Using getTargetRow() after append() is UB!
+   *
+   *  @param row(in)  input row
+   *  @param rowOut() row to aggregate data from input row
+   *
+   *  @returns true if new row created, false otherwise
+   */
+  bool getTargetRow(const Row& row, Row& rowOut);
+
+  /** @brief Dump some RGDatas to disk and release memory for further use.
+   */
+  void dump();
+
+  /** @brief Append RGData from other RowAggStorage and clear it.
+   *
+   *    NB! Any operation except getNextRGData() or append() is UB!
+   *
+   * @param other(in) donor storage
+   */
+  void append(RowAggStorage& other);
+
+  /** @brief Remove last RGData from internal RGData storage and return it.
+   *
+   * @returns pointer to the next RGData or nullptr if empty
+   */
+  std::unique_ptr<RGData> getNextRGData();
+
+  /** @brief TODO
+   *
+   * @param mergeFunc
+   * @param rowOut
+   */
+  void finalize(std::function<void(Row &)> mergeFunc, Row &rowOut);
+
+private:
+  /** @brief Create new RowAggStorage with the same params and load dumped data
+   *
+   * @param gen(in) generation number
+   * @return pointer to a new RowAggStorage
+   */
+  RowAggStorage* clone(uint16_t gen) const;
+
+  /** @brief Free any internal data
+   */
+  void freeData();
+
+  /** @brief Move internal data & row position inside [insIdx, startIdx] up by 1.
+   *
+   * @param startIdx(in) last element's index to move
+   * @param insIdx(in)   first element's index to move
+   */
+  void shiftUp(size_t startIdx, size_t insIdx);
+
+  inline static size_t mixHash(size_t x) noexcept
```

Comment:
I am curios which hash function mixer did you re-use or it was your own and why did you choose uint64_t and not uint32_t as the hash output?

Reply:
mixHash was left over after using the original Row::hash hasher, I removed it. robinhood hash looks more evenly distributed.
And 64 bits for hash was chosen simply because RH uses 64 bits.